Tasks/JavaToolInstaller/javatoolinstaller.ts

```diff
@@ -0,0 +1,108 @@
+    try {
+        let versionSpec = taskLib.getInput('versionSpec', true);
+        await getJava(versionSpec);
+    }
+    catch (error) {
+        console.error('ERR:' + error.message);
+    } 
+}
+
+async function getJava(versionSpec: string) {
+    const fromAzure: boolean = ('AzureStorage' == taskLib.getInput('jdkSource', true));
+    const extractLocation: string = taskLib.getPathInput('jdkDestinationDirectory', true);
+    const cleanDestinationFolder: boolean = taskLib.getBoolInput('cleanDestinationFolder', false);
+    let compressedFileExtension: string;
+    let jdkDirectory: string;
+
+    toolLib.debug('Trying to get tool from local cache first');
+    const localVersions: string[] = toolLib.findLocalToolVersions('Java');
+    const version: string = toolLib.evaluateVersions(localVersions, versionSpec);
+
+     // Clean the destination folder before downloading and extracting?
+     if (cleanDestinationFolder && taskLib.exist(extractLocation) && taskLib.stats(extractLocation).isDirectory) {
+        console.log(taskLib.loc('CleanDestDir', extractLocation));
+
+        // delete the contents of the destination directory but leave the directory in place
+        fs.readdirSync(extractLocation)
+        .forEach((item: string) => {
+            let itemPath = path.join(extractLocation, item);
+            taskLib.rmRF(itemPath);
+        });
+    }
+
+    if (version) { //This version of Java JDK is already in the cache. Use it instead of downloading again.
+        console.log(taskLib.loc("Info_ResolvedToolFromCache", version));
+    }
+    else if (fromAzure) { //Download JDK from an Azure blob storage location and extract.
+        try {
+            compressedFileExtension = getFileEnding(taskLib.getInput('azureCommonVirtualPath', true));
+        
+            let azureDownloader = new azureStorageDownloader.AzureStorageArtifactDownloader(taskLib.getInput('azureResourceManagerEndpoint', true), taskLib.getInput('azureStorageAccountName', true), taskLib.getInput('azureContainerName', true), taskLib.getInput('azureCommonVirtualPath', false));
+            await azureDownloader.downloadArtifacts(extractLocation, '*' + compressedFileExtension);
+            await sleepFor(250); //Wait for the file to be released before extracting it.
+
+            let extractSource = buildFilePath(extractLocation, compressedFileExtension);
+            jdkDirectory = new JavaFilesExtractor().unzipJavaDownload(extractSource, compressedFileExtension, extractLocation);
+        } catch (err) {
+            taskLib.error(err.message);
+            taskLib.setResult(taskLib.TaskResult.Failed, err.message);
+        }
+    }
+    else { //JDK is in a local directory. Extract to specified target directory.
+        try{
+            compressedFileExtension = getFileEnding(taskLib.getInput('jdkPath', true));
+            jdkDirectory = new JavaFilesExtractor().unzipJavaDownload(taskLib.getInput('jdkPath', true), compressedFileExtension, extractLocation);
```

Comment:
What should happen in the case that it doesn't have a file extension that we support? (For instance `.rar`?  Or in the case that they misconfigured it?)

Reply:
Good point. In my initial PR I had it such that the user specified the file type from a pickLIst. I removed that per one of the review suggestions but missed adding in additional logic to catch an unsupported extension.